src/cargo/util/network.rs

```diff
@@ -0,0 +1,88 @@
+use util::{CargoResult, Config, errors};
+use term::color::YELLOW;
+
+/// Wrapper method for network call retry logic.
+///
+/// Retry counts provided by Config object 'net.retry'. Config shell outputs
+/// a warning on per retry.
+///
+/// Closure must return a CargoResult.
+///
+/// Example:
+/// use util::network;
+/// cargo_result = network.with_retry(&config, || something.download());
+pub fn with_retry<T, E, F>(config: &Config, mut callback: F) -> CargoResult<T>
+    where F: FnMut() -> Result<T, E>,
+          E: errors::NetworkError
+{
+    let mut remaining = try!(config.net_retry());
+    loop {
+        match callback() {
+            Ok(ret) => return Ok(ret),
+            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
+                try!(config.shell()
+                           .say(format!("failed network call. {} \
+                                        tries remaining", remaining),
+                                YELLOW));
```

Comment:
Could this use `shell().warn(..)`? It also may want to avoid using a period as we tend to avoid that kind of punctuation in Cargo.